### PR TITLE
Feature/edit profile tutor

### PR DESCRIPTION
**Pull request by aferrante1**
### **Description:**  
This pull request introduces two screens, EditTutorProfile and EditTutorSchedule, allowing users to update their information to enhance their matching opportunities.

### **Key Changes:**  
- Created EditTutorProfile.kt and EditTutorSchedule.kt, two new screens for tutors to interact with.
- Added the corresponding navigation in MainActivity to facilitate easy access to these screens.
- Updated variable initialization to manage the changes in the tutor's information more effectively.
- Integrated the EditSchedule screen into NavigationActions for streamlined navigation.

### **Motivation:**  
The motivation behind this implementation is to provide tutors with the ability to easily manage and update their profiles and schedules. This feature enhances the user experience by ensuring that tutors can present accurate and up-to-date information, which is crucial for improving their matching opportunities with students. By allowing seamless updates to their profiles and schedules, we aim to increase tutor engagement and satisfaction within the app.

### **Affected Files:**
- MainActivity: Added the composable screens to the navigation.
- Profile.kt: Initialized the variables differently to update the values.
- NavigationActions: Added the EditSchedule screen.
- EditTutorProfile.kt: New screen for editing tutor profiles.
- EditTutorSchedule.kt: New screen for editing tutor schedules.
- ProfileInfo.kt: Added navigation to the new corresponding screens.

### CODE REVIEW

File: app/src/main/java/com/github/se/project/ui/profile/EditTutorProfile.kt [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+) {
+  val profile =
+      listProfilesViewModel.currentProfile.collectAsState().value
+          ?: return Text(
+              text = "No Profile selected. Should not happen.",
+              modifier = Modifier.testTag("editTutorNoProfile"))
+
+  val initialLanguagesList: List<Language> = profile.languages
+  val profileLanguages: SnapshotStateList<Language> = remember {
+    mutableStateListOf(*initialLanguagesList.toTypedArray())
```
Comment (alexs-rufer) [inline app/src/main/java/com/github/se/project/ui/profile/EditTutorProfile.kt:42]: I am actually really struggling to understand that pointer reference..

File: app/src/main/java/com/github/se/project/ui/profile/EditTutorProfile.kt [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+                  "Tutoring price per hour:",
+                  style = MaterialTheme.typography.titleSmall,
+                  modifier = Modifier.testTag("editTutorProfilePriceText"))
+              PriceSlider(priceSliderValue)
+
+              Spacer(modifier = Modifier.height(8.dp))
+
+              Text(text = "Modify your section", style = MaterialTheme.typography.titleSmall)
+              // Section dropdown menu with improved styling
+              Box {
```
Comment (alexs-rufer) [inline app/src/main/java/com/github/se/project/ui/profile/EditTutorProfile.kt:123]: we should be able to extract that Section selector (line 123 to 152) into a composable since this is the exact same code as in the CreateUserProfile.kt. I'll do it later its fine for now!

File: app/src/main/java/com/github/se/project/ui/profile/EditTutorProfile.kt [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+                              section.value = s.name
+                              expandedSection = false
+                            },
+                            modifier =
+                                Modifier.testTag("editTutorProfileSectionDropdownItem-${s.name}"))
+                      }
+                    }
+              }
+
+              Text(text = "Modify your academic level", style = MaterialTheme.typography.titleSmall)
```
Comment (alexs-rufer) [inline app/src/main/java/com/github/se/project/ui/profile/EditTutorProfile.kt:154]: Same comment here

File: app/src/main/java/com/github/se/project/ui/profile/EditTutorSchedule.kt [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+      listProfilesViewModel.currentProfile.collectAsState().value
+          ?: return Text(
+              text = "No Profile selected. Should not happen.",
+              color = Color.Red,
+              modifier = Modifier.testTag("editScheduleNoProfile"))
+
+  var profileSchedule by remember { mutableStateOf(profile.schedule) }
+  val context = LocalContext.current
+
+  // prnt the profile schedule
```
Comment (alexs-rufer) [inline app/src/main/java/com/github/se/project/ui/profile/EditTutorSchedule.kt:37]: We should remove that print. Maybe it was for debugging purpose

**Review by alexs-rufer — COMMENTED** [review]
These are some great screens, thanks for the work :) I've added a few suggestions and I'll work on that bug!

File: app/src/main/java/com/github/se/project/ui/profile/EditTutorProfile.kt [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+        viewModel(factory = ListProfilesViewModel.Factory)
+) {
+  val profile =
+      listProfilesViewModel.currentProfile.collectAsState().value
+          ?: return Text(
+              text = "No Profile selected. Should not happen.",
+              modifier = Modifier.testTag("editTutorNoProfile"))
+
+  val initialLanguagesList: List<Language> = profile.languages
+  val profileLanguages: SnapshotStateList<Language> = remember {
```
[the remark this answers is not in the document]
Comment (aferrante1) [reply app/src/main/java/com/github/se/project/ui/profile/EditTutorProfile.kt:41]: right I forgot to switch them up! Thank you!

File: app/src/main/java/com/github/se/project/ui/profile/EditTutorProfile.kt (hunk 1 above)
Comment (aferrante1) [reply app/src/main/java/com/github/se/project/ui/profile/EditTutorProfile.kt:42]: As before we had the enum i had to work around it, but now that we have switched to Listof I need to initialise the variables differently

File: app/src/main/java/com/github/se/project/model/profile/Profile.kt [hunk 6]
```diff
@@ -8,10 +8,10 @@ data class Profile(
...
-    val section: Section, // Section of the user
-    val academicLevel: AcademicLevel, // Academic level of the user
-    val languages: List<Language> = listOf(),
-    val subjects: List<Subject> = listOf(),
-    val schedule: List<List<Int>> = List(7) { List(12) { 0 } }, // Weekly schedule
+    var section: Section, // Section of the user
+    var academicLevel: AcademicLevel, // Academic level of the user
+    var languages: List<Language> = listOf(),
+    var subjects: List<Subject> = listOf(),
+    var schedule: List<List<Int>> = List(7) { List(12) { 0 } }, // Weekly schedule
```
[the remark this answers is not in the document]
Comment (aferrante1) [reply app/src/main/java/com/github/se/project/model/profile/Profile.kt:15]: I found a solution we can use Launched effect to populate the variable. If you print for debugging purposes you find out they are empty once initialized.

File: app/src/androidTest/java/com/github/se/project/ui/profile/EditTutorProfileTest.kt [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+    composeTestRule.onNodeWithTag("editTutorProfileSectionDropdown").performClick()
+    composeTestRule
+        .onNodeWithTag("editTutorProfileSectionDropdownItem-GM")
+        .assertIsDisplayed()
+        .performClick()
+
+    composeTestRule.onNodeWithTag("editTutorProfileConfirmButton").performClick()
+
+    assertEquals(AcademicLevel.MA4, mockViewModel.currentProfile.value?.academicLevel)
+    assertEquals(Section.GM, mockViewModel.currentProfile.value?.section)
```
Comment (A-M-00) [inline app/src/androidTest/java/com/github/se/project/ui/profile/EditTutorProfileTest.kt:111]: Could be useful to have an equivalent test that also edits languages/subjects/name, but otherwise good job

File: app/src/androidTest/java/com/github/se/project/ui/profile/EditTutorProfileTest.kt [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+            section = Section.IN,
+            academicLevel = AcademicLevel.MA2,
+            schedule = List(7) { List(12) { 0 } })
+    // Set the screen in the test environment
+    composeTestRule.setContent {
+      EditTutorProfile(navigationActions = mockNavigationActions, mockViewModel)
+    }
+    composeTestRule.onNodeWithTag("editTutorProfileConfirmButton").performClick()
+
+    Mockito.verify(mockNavigationActions, Mockito.never()).navigateTo(Mockito.anyString())
```
Comment (A-M-00) [inline app/src/androidTest/java/com/github/se/project/ui/profile/EditTutorProfileTest.kt:155]: Same as above, could be interesting to test other cases. As is this works though.

File: app/src/androidTest/java/com/github/se/project/ui/profile/EditTutorProfileTest.kt [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+      EditTutorProfile(
+          navigationActions = mockNavigationActions,
+          listProfilesViewModel = viewModel(factory = ListProfilesViewModel.Factory))
+    }
+
+    composeTestRule
+        .onNodeWithTag("editTutorNoProfile")
+        .assertTextEquals("No Profile selected. Should not happen.")
+  }
+}
```
Comment (A-M-00) [inline app/src/androidTest/java/com/github/se/project/ui/profile/EditTutorProfileTest.kt:172]: Good edge-case test!

File: app/src/androidTest/java/com/github/se/project/ui/profile/EditTutorScheduleTest.kt [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+    composeTestRule
+        .onNodeWithTag("editScheduleInstructionsText")
+        .assertTextEquals("Modify the time slots you're available during the week:")
+
+    // Verify that the "Let's find a student!" button is displayed and has the correct text
+    composeTestRule.onNodeWithTag("editScheduleButton").assertTextEquals("Update Schedule")
+    composeTestRule.onNodeWithTag("editScheduleButton").performClick()
+
+    // Verify that the updateProfile was called
+    // assertEquals(AcademicLevel.MA4, mockViewModel.currentProfile.value?.academicLevel)
```
Comment (A-M-00) [inline app/src/androidTest/java/com/github/se/project/ui/profile/EditTutorScheduleTest.kt:80]: These should probably be deleted

File: app/src/androidTest/java/com/github/se/project/ui/profile/EditTutorScheduleTest.kt
Comment (A-M-00) [inline app/src/androidTest/java/com/github/se/project/ui/profile/EditTutorScheduleTest.kt:1]: The tests here are well done, but a test that measures whether profile editing happens successfully should probably be added

File: app/src/main/java/com/github/se/project/MainActivity.kt
Comment (A-M-00) [inline app/src/main/java/com/github/se/project/MainActivity.kt:1]: Nice to see navigation already implemented!

File: app/src/main/java/com/github/se/project/ui/profile/EditTutorProfile.kt [hunk 11]
```diff
@@ -0,0 +1,221 @@
...
+fun EditTutorProfile(
+    navigationActions: NavigationActions,
+    listProfilesViewModel: ListProfilesViewModel =
+        viewModel(factory = ListProfilesViewModel.Factory)
+) {
+  val profile =
+      listProfilesViewModel.currentProfile.collectAsState().value
+          ?: return Text(
+              text = "No Profile selected. Should not happen.",
+              modifier = Modifier.testTag("editTutorNoProfile"))
```
Comment (A-M-00) [inline app/src/main/java/com/github/se/project/ui/profile/EditTutorProfile.kt:37]: Good edge case handling

File: app/src/main/java/com/github/se/project/ui/profile/EditTutorProfile.kt [hunk 12]
```diff
@@ -0,0 +1,221 @@
...
+  val profileLanguages = remember { mutableStateListOf<Language>() }
+  val profileSubjects = remember { mutableStateListOf<Subject>() }
+
+  LaunchedEffect(profile) {
+    profileLanguages.clear()
+    profileLanguages.addAll(profile.languages)
+
+    profileSubjects.clear()
+    profileSubjects.addAll(profile.subjects)
+  }
```
Comment (A-M-00) [inline app/src/main/java/com/github/se/project/ui/profile/EditTutorProfile.kt:48]: Elegant solution!

File: app/src/main/java/com/github/se/project/ui/profile/EditTutorProfile.kt [hunk 13]
```diff
@@ -0,0 +1,221 @@
...
+                    .padding(horizontal = 16.dp)
+                    .padding(paddingValues)
+                    .testTag("tutorInfoScreen"),
+            verticalArrangement = Arrangement.spacedBy(8.dp)) {
+              Text(
+                  text = "${profile.firstName} ${profile.lastName}",
+                  modifier =
+                      Modifier.testTag(
+                          "editTutorWelcomeText"), // .padding(vertical = 48.dp, horizontal =
+                  // 16.dp).testTag("welcomeText"),
```
Comment (A-M-00) [inline app/src/main/java/com/github/se/project/ui/profile/EditTutorProfile.kt:83]: Ideally, comments should be removed later

File: app/src/main/java/com/github/se/project/ui/profile/EditTutorProfile.kt [hunk 14]
```diff
@@ -0,0 +1,221 @@
...
+              LanguageSelector(profileLanguages)
+
+              Spacer(modifier = Modifier.height(16.dp))
+
+              // Subject Selection
+              Text(
+                  "Teaching subjects:",
+                  style = MaterialTheme.typography.titleSmall,
+                  modifier = Modifier.testTag("editTutorProfileSubjectText"))
+              SubjectSelector(null, profileSubjects, true)
```
Comment (A-M-00) [inline app/src/main/java/com/github/se/project/ui/profile/EditTutorProfile.kt:108]: Good initiative, using the pre-existing LanguageSelector and SubjectSelector components

File: app/src/main/java/com/github/se/project/ui/profile/EditTutorProfile.kt
Comment (A-M-00) [inline app/src/main/java/com/github/se/project/ui/profile/EditTutorProfile.kt:1]: Ideally, there should also be a possibility to edit phone number/name. This is great for now though, I'll add the others later

File: app/src/main/java/com/github/se/project/ui/profile/EditTutorSchedule.kt [hunk 15]
```diff
@@ -0,0 +1,86 @@
...
+                  "Modify the time slots you're available during the week:",
+                  style = MaterialTheme.typography.bodyLarge,
+                  modifier = Modifier.testTag("editScheduleInstructionsText"))
+              Spacer(modifier = Modifier.height(8.dp))
+
+              AvailabilityGrid(
+                  schedule = profileSchedule,
+                  onScheduleChange = { updatedSchedule -> profileSchedule = updatedSchedule },
+                  modifier = Modifier.weight(1f))
+            }
```
Comment (A-M-00) [inline app/src/main/java/com/github/se/project/ui/profile/EditTutorSchedule.kt:71]: Again, good use of components

File: app/src/main/java/com/github/se/project/ui/profile/ProfileInfo.kt [hunk 16]
```diff
@@ -50,12 +51,12 @@ fun ProfileInfoScreen(
             horizontalArrangement = Arrangement.SpaceBetween) {
               Row(Modifier.padding(horizontal = 4.dp)) {
                 IconButton(
-                    onClick = { /* Navigate to calendar screen */},
+                    onClick = { navigationActions.navigateTo(Screen.EDIT_SCHEDULE) },
```
Comment (A-M-00) [inline app/src/main/java/com/github/se/project/ui/profile/ProfileInfo.kt:54]: Should have an if condition to navigate to the student equivalent, but again can be added later

**Review by A-M-00 — APPROVED** [review]
Overall, great work! Should improve the app significantly.
